### ISSUE TITLE
Proposal: Volatile captured variables

### ISSUE THREAD

**Issue by acelent** [dotnet/roslyn#2234]
The `volatile` keyword can be just as useful on a captured local variables as it is on fields.

For instance, in an iterator, in an async function or in an anonymous function passed to `System.Threading.Thread`'s constructor, you may need to read the current value of a captured variable, instead of some cached, possibly stale value.

Before C# 2.0, it was true that [locals were not *address exposed*](http://blogs.msdn.com/b/cbrumme/archive/2003/05/17/51445.aspx), thus they could not be accessed from multiple threads in parallel.  But since C# 2.0, locals are only truly local if they're not captured and are not inside an iterator, and are not inside an async function since C# 5.0.

Depending on implementation details, the captured variables may end up being fields, in which case the compiler can just transfer the `volatile` declaration and continue to do its thing.  If the variables end up somewhere else (e.g. an array, in case they're all of the same type), the compiler can use `System.Threading.Thread.VolatileRead`/`System.Threading.Thread.VolatileWrite` or `System.Threading.Volatile.Read`/`System.Threading.Volatile.Write` methods to achieve the same effect.

In the case of iterators and async functions, since they're compiled into state machines, the variables declared in them may also be declared `volatile`.  In the case of anonymous async functions, every variable it captures or declared may be `volatile`.

The purpose of this proposal is code simplification, in order to allow `volatile` semantics without having to define another class.

**Comment by HaloFour** [comment]
In the case of the state machines for async and iterators aren't those inherently not thread-safe?  (I know that sounds weird with async but it's not like the continuations are executing in parallel.)

As for standard closures I can see where this could potentially be a problem.  I believe that volatile access adds overhead so it would probably be preferable to opt-in perhaps by marking the variable as `volatile`.  The compiler can then either fail if the variable is not lifted or silently ignore it.

**Comment by VSadov** [comment]
This sounds like a request to allow volatile modifier on local variables and formal parameters - just in case they would be implemented as fields because of a capture in a lambda, iterator, async or something like that.

It seems that the scenarios where that would be useful are much less common than scenarios where volatile would have no meaning. 
Most locals/parameters are no captured and it would be strange to limit the applicability only to captured ones. Capturing is essentially an implementation detail that programmer can control only very indirectly. For example async often captures more variables in Debug than in Release as a choice between debuggability and performance.

So far it looks like this scenario is better served by Volatile.Read / Volatile.Write

**Comment by acelent** [comment]
@HaloFour, you're right that iterators don't have to support parallel execution, and async continuations don't run in parallel.  But the issue is about stale data.

That said, an iterator can move in different threads, in one thread at a time, and async continuations can run in different threads (e.g. thread pool), so in any case you risk having stale data being read.

One case where this is important is when flagging something to the iterator or async function from another thread (e.g. a background thread).

As for the opt-in, I totally agree.

**Comment by acelent** [comment]
@VSadov, by that line of thought, the `volatile` keyword isn't needed even for fields.

Note that I mention that capturing variables into fields is an implementation detail.  It could also use an array, or some other internal mechanism, but `volatile` would still be useful no matter how it would be implemented.

Having supposed an implementation was just a way to transmit the line of though, the reason behind the usefulness or even the need for `volatile` for such variables.

As for supporting only captured variables, I guess actual locals could be allowed to be declared `volatile`, being silently ignored by the compiler.  I restrained the proposal by that much because that's the extent on which `volatile` has a meaning.  We can leave to a code style checker to warn the developer that `volatile` won't do nothing for actual locals.

**Comment by svick** [comment]
@acelent 

> One case where this is important is when flagging something to the iterator or async function from another thread (e.g. a background thread).

I have a hard time imagining how you would do this without using fields or lambdas in a way that's not convoluted. Could you give an example?

**Comment by acelent** [comment]
@svick, the iterator or async function can have a `bool` variable at the top which can be set through various stages (e.g. inside a loop), e.g. `finished` or `isLast`.

Suppose the meaning of this flag is so that you can `yield` or `await` now, but return a different value or throw an exception the next time by checking the flag and using the values of the other non-volatile variables.  Although an iterator or an async continuation isn't supposed to run in parallel, it can run in another thread other than the initial/previous one.  As such, it should be possible to observe the changes to all other non-volatile changed variables before the write to a `volatile` variable.

Right now, you need to define a class and create an object with a `volatile` field, or use `Thread.VolatileRead`/`Thread.VolatileWrite` or `Volatile.Read`/`Volatile.Write` everywhere the supposedly volatile variable is used.  The former takes care of compiling to volatile reads and writes but requires (further) heap allocation, the latter is prone to human error where one can easily miss a volatile call.

Also, it's an implementation detail that .NET has release semantics on writes/stores since 2.0 ([source 1](https://msdn.microsoft.com/en-us/magazine/cc163715.aspx), [source 2](http://joeduffyblog.com/2007/11/10/clr-20-memory-model/)), so this *usually* works without volatile variables.  My proposal is in the vein of not depending on this implementation detail.

Nonetheless, captured variables will always benefit from being able to be declared `volatile`, because the compiler itself will avoid stale reads.

**Comment by svick** [comment]
@acelent Are you saying that if you have code like this:

```c#
bool b = true; // 1
await someTask;
b = false; // 2
await anotherTask;
Console.WriteLine(b); // 3
```

Then if lines 1 and 3 ran on the same thread and line 2 ran on another thread, are you saying that `b` is not guaranteed to be observed as `false`? If that were true, then pretty much every single `async` method would be broken, I think.

I couldn't find this properly documented anywhere, except for [a comment](http://stackoverflow.com/a/13361366/41071) by @stephentoub, which says that barriers are added automatically, so there should be no need for `volatile` in `async` methods without lambdas (not sure about iterators).

**Comment by VSadov** [comment]
@acelent 
>  by that line of thought, the volatile keyword isn't needed even for fields.

Generally speaking yes. Volatile isn't needed even for fields. 

The landscape of computing has changed since late 90s when volatile was introduced, We have deeper understanding of concurrent programming and what patterns work better than others. I have doubts that volatile fields would make it into the language if C# 1.0 was designed now.

Generally it is better to specify memory ordering at the places where memory is accessed. 
Specifying access mode at the level of variables is problematic because:
1) specifier is too far from its effects - volatile is applied at declarations, but has effect at accesses. That requires extra effort to remember which variables are volatile and which are not when looking at the code. That makes it harder to reason about correctness.
2) heavy-handled - now all accesses to the data represented by that variable are volatile, even when they do not need to be (like if accesses are already synchronized by running under a lock or in an async method). 
I.E if you do not care about costs of synchronization, locks are easier to reason about and thus are generally safer. If you have to use volatile, you are likely implementing a perf sensitive low-lock scenario and reducing unnecessary synchronization is your main goal. Unconditional synchronization must be used very carefully in such case.
3) incomplete - not all locations in memory can be decorated volatile. - like array elements
4) deceivingly, volatile does not actually apply to the data location. - volatile is a property of a variable (in lexical sense), which is insufficient when same memory location is accessed through multiple variables. For example accessing same location via byref parameter will not know if the location is referenced via volatile field and such accesses will not be automatically volatile.

Basically, in cases of any complexity above trivial, it is often more straightforward (and safe) to just use Volatile.XXXX, rather than prove that volatile, with all its subtleties, results in expected behavior.

For another, perhaps more radical, look at the same problem you may read: http://joeduffyblog.com/2010/12/04/sayonara-volatile/

In my view (!), decorating _accesses_ as volatile might be interesting, but it is hard to add much value here since Volatile.Read/Volatile.Write is quite short and expressive way to do that already.
As for expanding the applicability of volatile to more variable kinds - I am not sure it is a very rewarding direction for the language.

**Comment by acelent** [comment]
@svick, it so happens that (de)queueing or "stealing" a task will probably use a lock, an interchange operation or an explicit memory barrier, so I guess it's implicit.

@VSadov, I kind of agree, but `volatile` is here, it could just as well be more consistent about which identifiers can be `volatile`.  I'd fully agree if we were to instead remove `volatile` from the language entirely.

I see what you mean with 1), but 1) also means you don't have to worry about surrounding the variable with a volatile accessor method.  Which brings us to 2), it doesn't make any sense at all to use a volatile place without volatile semantics.  If you want that, read the volatile place into a non-volatile place (e.g. a `volatile` field into a non-volatile variable).

As for 3), that's perhaps one of the main purposes of the volatile accessor methods, as you can pass them a reference to an array element.  I think it would be cumbersome to define syntax that says "treat array elements of this place as `volatile`".  Or maybe not, by allowing the `volatile` keyword inside the square brackets of an array variable type specifier, e.g. `int[volatile] args;`, but this is material for a different proposal.

Finally, 4) is perfectly fine for me.  In fact, you kind of contradict yourself in saying in 2) that always having volatile accesses on the same place (e.g. field, variable) is a burden, but it would be desirable to have volatility on the actual location (e.g. memory location).  If you had on the location, you'd eliminate a currently allowed compiler optimization of removing adjacent reads or adjacent writes, since the compiler wouldn't know whether a location was volatile, thus whether the side-effects had to be visible from/to other threads.

But I think this already happens for writes right now for Microsoft .NET anyway, it's just not standardized, so it's implementation dependent, e.g. Mono can have weaker write semantics.

The fact that you lose volatile semantics when passing by ref is not much of a problem, code analysis done by the compiler already warns you of such cases.  In fact, perhaps `ref` (and `out`) parameters should be allowed to be `volatile`, but that's yet material for a different proposal.

**Comment by VSadov** [comment]
4 and 2 are not really contradicting. 
Point 2 shows that volatile can easily "overdo" by introducing synchronization that is not apparent, while 4 shows that there is still a way to "underdo". Overdoing in some scenarios and underdoing in others is not contradictory Both are just things that a programmer must keep in mind when using the feature and are not good for the feature's usability.

Removing the feature is not an option at this point, since there is definitely enough code already written that through careful design and testing uses this feature and behaves as intended.

Not removing a feature, however, is rarely by itself a reason to invest more in it. There must be a scenario that is not well served by existing features (having Volatile accessors, makes this part doubtful) and it would be nice if the new feature would play well with existing features and have as few special cases or behaviors as possible. 

Just consider how many places where locals can be introduced - regular locals, for/foreach loops, catch clauses, using, fixed, lock statements and such, query comprehensions ... 
Should all these allow volatile, or just some of them? Would it be worth it for a relatively rare scenario that already has a fairly decent solution?

**Comment by acelent** [comment]
@VSadov, the volatile declaration can be far from the volatile accesses as any other declaration: type, visibility, static, readonly, etc.  For instance, worse than volatile is that you must use conventions to distinguish a field from a property.

So, 2) does something very useful for the developer, which is to make sure every access is volatile (not just writes, as it is in Microsoft .NET but not in the ECMA), and 4) is a consequence of volatile accesses being bound to the declared field (or as proposed, declared variable).  I don't know if you agree, but there's no sense in using volatile **reads** to a field (or variable) in one piece of code and not in other pieces of code.

For instance, volatile write semantics by default, as in Microsoft .NET, is not enough if some other thread is using non-volatile reads from the same field: within the same method, the .NET language compiler is allowed to merge adjacent reads, and the native architecture may go a step further if it has a *weak* memory model.

It's much easier to "underdo" by having to use a volatile accessor method than it is to "overdo" by having a field (or variable) declared volatile.  And really, even if you "overdo", you lose a bit of performance, but not even a bit of correctness, whereas "underdoing" is really dangerous.

As for the many places where this may be applicable, if it is such a hurdle to make the change for all of them, then the code needs be more modular.  I thought locals are the same everywhere, and that in Microsoft .NET, captured variables, iterator variables and async method variables are fields.

In captured variables, iterator variables and async method variables, it would simply let the `volatile` declaration go through in the code transformation and the compiler would do its thing (prefix accesses with CIL `volatile.`).  This is what motivated me to create the proposal in the first place, it's (apparently) a simple thing.

As for actual local variables, I don't really care if the compiler ignores or warns the volatile declaration, although I think it shouldn't error.

PS: One perhaps far-fetched example: if you unit test a non-thread-bound async method by:

1. Create two threads
1. In one thread, call the async method and get the returned task
3. In the other thread, wait for the task result, forcing its execution

If you manage to do this without using barriers, you can get stale values.

As for an iterator used in multiple threads (interleaved), you can get a stale value perhaps more easily than with async methods.

---
Finally, what I **really** care about is to be able to declare `volatile` variables in anonymous methods (or lambdas).

I want to emphasize this by stating that, even though I included iterator methods and async methods simply because I thought it would be just as simple to add such feature, I'm willing to make that part of the proposal optional, i.e. in practice give up on it.

**Comment by gafter** [comment]
We've decided not to do this. There are other ways to accomplish what this would do, and it doesn't pay for itself.

**Comment by acelent** [comment]
I still have hope that #49, in case it's accepted, also considers allowing the `volatile` modifier in either static or plain method variables.

**Comment by acelent** [comment]
MSDN has an example in the [`SpinWait` structure documentation](https://msdn.microsoft.com/en-us/library/system.threading.spinwait(v=vs.110).aspx) which may loop indefinitely, in case an implementation doesn't perform regular writes as volatile writes (watch the variable `someBoolean`):

```c#
using System;
using System.Threading;
using System.Threading.Tasks;

class SpinWaitDemo
{
    // Demonstrates:
    //      SpinWait construction
    //      SpinWait.SpinOnce()
    //      SpinWait.NextSpinWillYield
    //      SpinWait.Count
    static void Main()
    {
        bool someBoolean = false;
        int numYields = 0;

        // First task: SpinWait until someBoolean is set to true
        Task t1 = Task.Factory.StartNew(() =>
        {
            SpinWait sw = new SpinWait();
            while (!someBoolean)
            {
                // The NextSpinWillYield property returns true if
                // calling sw.SpinOnce() will result in yielding the
                // processor instead of simply spinning.
                if (sw.NextSpinWillYield) numYields++;
                sw.SpinOnce();
            }

            // As of .NET Framework 4: After some initial spinning, SpinWait.SpinOnce() will yield every time.
            Console.WriteLine("SpinWait called {0} times, yielded {1} times", sw.Count, numYields);
        });

        // Second task: Wait 100ms, then set someBoolean to true
        Task t2 = Task.Factory.StartNew(() =>
        {
            Thread.Sleep(100);
            someBoolean = true;
        });

        // Wait for tasks to complete
        Task.WaitAll(t1, t2);
    }
}
```

Although this is just an example, it is not far fetched at all to have more complicated behavior with delegates or lambdas and with more variables that should always be accessed as volatile.

**Comment by acelent** [comment]
In fact, the previous example may enter an infinite loop even if the jitter issues volatile writes for normal writes, because it can read the variable `someBoolean` only once in the looping lambda, which may be `false` by the time it reads with high probability, given the order by which tasks are started..

If the variable could be marked as `volatile`, the compiler would know not to optimize accesses to it, guaranteeing that `someBoolean` would always be re-read.  Not to mention it would issue a volatile read, which would as well bring a read order guarantee after each time `someBoolean` has been read.
